transport/grpc/peer.go

```diff
@@ -57,132 +54,64 @@ func (t *Transport) newPeer(address string, options *dialOptions) (*grpcPeer, er
 		Peer:       hostport.NewPeer(hostport.PeerIdentifier(address), t),
 		t:          t,
+		ctx:        ctx,
+		cancel:     cancel,
 		clientConn: clientConn,
-		stoppingC:  make(chan struct{}),
 		stoppedC:   make(chan struct{}),
 	}
+
 	go grpcPeer.monitor()
+
 	return grpcPeer, nil
 }
 
 func (p *grpcPeer) monitor() {
-	defer p.monitorStop()
-	if !p.monitorStart() {
-		return
-	}
-
-	var attempts uint
-	backoff := p.t.options.backoffStrategy.Backoff()
-
-	connectivityState := p.clientConn.GetState()
-	changed := true
+	p.setStatus(peer.Unavailable)
+	var grpcStatus connectivity.State
 	for {
-		var peerConnectionStatus peer.ConnectionStatus
-		var err error
-		// will be called the first time since changed is initialized to true
-		if changed {
-			peerConnectionStatus, err = connectivityStateToPeerConnectionStatus(connectivityState)
-			if err != nil {
-				return
-			}
-			p.Peer.SetStatus(peerConnectionStatus)
-		}
+		grpcStatus = p.clientConn.GetState()
+		yarpcStatus := grpcStatusToYARPCStatus(grpcStatus)
+		p.setStatus(yarpcStatus)
 
-		var ctx context.Context
-		cancel := noop
-		if peerConnectionStatus == peer.Available {
-			attempts = 0
-			ctx = context.Background()
-		} else {
-			attempts++
-			ctx, cancel = context.WithTimeout(context.Background(), backoff.Duration(attempts))
+		if !p.clientConn.WaitForStateChange(p.ctx, grpcStatus) {
+			break
 		}
-
-		newConnectivityState, loop := p.monitorLoopWait(ctx, cancel, connectivityState)
-		if !loop {
-			return
-		}
-		changed = connectivityState != newConnectivityState
-		connectivityState = newConnectivityState
 	}
-}
+	p.setStatus(peer.Unavailable)
 
-// return true if the transport is started
-// return false is monitor was stopped in the meantime
-// this should only be called by monitor()
-func (p *grpcPeer) monitorStart() bool {
-	select {
-	// wait for start so we can be certain that we have a channel
-	case <-p.t.once.Started():
-		return true
-	case <-p.stoppingC:
-		return false
-	}
-}
-
-// this should only be called by monitor()
-func (p *grpcPeer) monitorStop() {
-	p.Peer.SetStatus(peer.Unavailable)
-	// Close always returns an error
+	// Close always returns an error.
 	_ = p.clientConn.Close()
 	close(p.stoppedC)
 }
 
-// this should only be called by monitor()
-// this does not correlate to wait() at all
-//
-// return true to continue looping
-func (p *grpcPeer) monitorLoopWait(ctx context.Context, cancel context.CancelFunc, connectivityState connectivity.State) (connectivity.State, bool) {
-	defer cancel()
-
-	changedC := make(chan bool, 1)
-	go func() { changedC <- p.clientConn.WaitForStateChange(ctx, connectivityState) }()
-
-	loop := false
-	select {
-	case changed := <-changedC:
-		if changed {
-			connectivityState = p.clientConn.GetState()
-		}
-		loop = true
-	case <-p.stoppingC:
-	case <-p.t.once.Stopping():
+func (p *grpcPeer) setStatus(status peer.ConnectionStatus) {
+	if p.status != status {
+		p.status = status
+		p.Peer.SetStatus(status)
```

Comment:
This seems odd, but it looks like we are trying to save the hostport peer from calling `NotifyStatusChanged` on its subscribers. Should this logic be in the `hostport.Peer` instead of the gRPC peer?